pkg/scheduler/schedule_one.go

```diff
@@ -62,6 +62,8 @@ const (
 // scheduleOne does the entire scheduling workflow for a single pod. It is serialized on the scheduling algorithm's host fitting.
 func (sched *Scheduler) scheduleOne(ctx context.Context) {
 	podInfo := sched.NextPod()
+	defer sched.DonePod(podInfo)
```

Comment:
(Continue the comment of keep signature of FailureHandlerFn and AddUnschedulableIfNotPresent here...)

So it makes more sense to ensure the in-flight Pod is tracked for both schedulingCycle AND bindingCycle: by the end the cycle's completion. And thus, we manage the in-flight Pod in a general way, and so maybe we don't need the special logic to deal with pod's invalidation in PreBind? Could you please verify if it's doable? If so, the delta changes would be pretty minimum and decent.

Reply:
I was worried that the logic around when to call `Done` would be worse, but it's not that bad: I decided to still have a defer at the top of `scheduleOne`, except that it now uses a variable that gets cleared once the goroutine takes over.

However, there is a snag: once it is time to modify the pod, `schedulingCycle` creates a deep copy of the entire `PodInfo` struct. Then `MoveAllToActiveOrBackoffQueue` updates `MoveRequestCycle` in the original instance whereas `FailureHandler` is called with the clone and therefore doesn't use the backoff queue.

To get around this, I hacked  `schedulingCycle` to do the deep copy only for the pod. When the same `PodInfo` instance is used for the pod, it works - or at least none of the tests fail. My worry is that this will cause problems in other scenarios, like for example when receiving pod update events at exactly the wrong time?

I've pushed a revert of the additional parameter commit, a simple fixup for the review nits and the `Done` change.

Reply:
An alternative that keeps the `assumedPodInfo := podInfo.DeepCopy()` in `schedulingCycle` would be to turn `inFlighPods` into a `map[types.UID]int` which maps `UID->MoveRequestCycle.` That's sufficient for the current purpose. But it sounds like you wanted to have more information about in flight pods?

Reply:
I've pushed one additional commit with that solution. This feels cleaner to me than the previous hack with not cloning PodInfo.

Reply:
I'm in favor of the latest version (`map[types.UID]cycleNumber`). Thanks!

Reply:
With the current `Done` parameter (= pointer) or shall I switch to a UID string?

Reply:
Never mind, I hadn't refreshed the GitHub page. I'll switch to UID string on Monday.